# Citation
If you use this resource, please cite:


Stalder, L., Maurhofer, M., & Croll, D. (2025). High-resolution profiling of bacterial and fungal communities using pangenome-informed taxon-specific amplicons and long-read sequencing. Microbiome.

# Content
01 - Scripts and files to create pangenomes

02 - Scripts and files to design primer on pangenome cores

03 - Scripts and files for PacBio raw read processing, as well as dada2 pipeline to create ASV tables

04 - Scripts and files for taxonomy assignments and phyloseq object creation for visualization

# Abstract
High-throughput sequencing technologies have greatly advanced our understanding of microbiomes, but resolving microbial communities at species and strain levels remains challenging. Here, we developed and validated a pipeline for designing, multiplexing, and sequencing highly polymorphic taxon-specific long-read amplicons. We focused on the wheat microbiome as a proof-of-principle and demonstrate strain-level resolution for the wheat-associated Pseudomonas microbiome and the ubiquitous fungal pathogen Zymoseptoria tritici. We achieved an order of magnitude higher phylogenetic resolution compared to existing ribosomal amplicons. The designed amplicons accurately capture species and strain diversity outperforming full-length 16S and ITS amplicons. Furthermore, we tracked microbial communities in the wheat phyllosphere across time and space to establish fine-grained species and strain-specific dynamics. To expand the utility of our approach, we generated pangenome-informed amplicon templates for additional key bacterial and fungal genera. Pangenome-informed microbiome profiling enables the tracking of microbial community dynamics in complex environments and overcomes limitations in phylogenetic resolution.
